File: 1514. Path with Maximum Probability.py
# 1514. Path with Maximum Probability


class Solution:
    def maxProbability(
        self,
        n: int,
        edges: List[List[int]],
        succProb: List[float],
        start_node: int,
        end_node: int,
    ) -> float:
        neimap = {i: [] for i in range(n)}
        for i in range(len(succProb)):
            neimap[edges[i][0]].append((-succProb[i], edges[i][1]))
            neimap[edges[i][1]].append((-succProb[i], edges[i][0]))

        visit = set()
        hp = []
        for start in neimap[start_node]:
            heappush(hp, start)
        ans = -1.0

        while hp:
            prob, node = heappop(hp)
            if node == end_node:
                ans = max(-prob, ans)
                return ans
            else:
                visit.add(node)
            for nei_prob, nei in neimap[node]:
                if nei in visit:
                    continue
                heappush(hp, (-nei_prob * prob, nei))
        # print(ans)
        return ans if ans != -1.0 else 0

File: 1514. Path with Maximum Probability.cpp
// 1514. Path with Maximum Probability


#include <vector>
#include <queue>
#include <unordered_map>
#include <functional>
#include <utility>
#include <limits>
using namespace std;



class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start_node, int end_node) {
        vector<vector<pair<int, double>>> adj(n);
        for (int i = 0; i < edges.size(); i++) {
            int n1 = edges[i][0], n2 = edges[i][1];
            double prob = succProb[i];
            adj[n1].emplace_back(n2, prob); // making (n1, prob) object directly and put it in to place vs push_back : already existing object
            adj[n2].emplace_back(n1, prob);
        }

        // max heap q - prob, node
        priority_queue<pair<double, int>> pq;
        pq.emplace(1.0, start_node);
        
        vector<double> maxProb(n, 0.0);
        maxProb[start_node] = 1.0;

        while (!pq.empty()) {
            auto [prob, node] = pq.top();
            pq.pop();
            
            if (prob < maxProb[node]) continue;
            
            for (auto& [nei, nei_prob] : adj[node]) {
                double new_prob = prob * nei_prob;
                if (new_prob > maxProb[nei]) {
                    maxProb[nei] = new_prob;
                    pq.emplace(new_prob, nei);
                }
            }
        }

        return maxProb[end_node];
    }
};